### PR TITLE
Tests relocations (from miner) with new testing tools

### PR DESCRIPTION
**Pull request by xandkar**
Mainly 2 things here:
1. relocation of 2 suites from miner:
  - the whole `miner_txn_bundle_SUITE` into `blockchain_txn_bundle_SUITE` and
  - selections from `miner_blockchain_SUITE` into `blockchain_vars_SUITE`;
2. new test-scenario setup tools:
  - `t_chain`: chain setup, block creation and addition
  - `t_user`: crypto key storage abstraction
  - `t_txn`: transaction creation shortcuts

But also some drive-by spec additions which I felt the need for when reading
the code.

While 1 was the main objective, 2 came out of the need to re-write the tests
innards anyway, so why not do it a little better? Also professor asked for
more-flexible scenario setup tools.

`t_chain` mainly improves on functionality of `test_utils:init*` and
`test_utils:*_block` functions by giving the user a lot more control over what
addresses end-up owning what assets - this makes testing scenarios a lot
clearer about what their base expectations are. Additionally, it adds the
`commit` shortcut, which bundles validation, block creation and addition to
chain - this was surprisingly absent from `test_utils`, but was needed for the
imported tests.

`t_user` is a key-holding abstraction which simplifies playing god and creating
scenarios with ownership transfers.

`t_txn` is just a place to organize transaction-construction shortcuts which
understand `t_user`, similar to `test_utils:create_payment_transaction`.

Using the new testing tools/abstractions, a payment test can look as intuitive
and transparent as:

```erlang
[{Src, SrcBalance} | _] = ?config(users_with_hnt, Cfg),
Dst = t_user:new(),
DstBalance = t_chain:get_balance(Chain, Dst),
Amount = 1000,
Txn = t_txn:pay(Src, Dst, Amount, 1),

?assertEqual(ok, t_chain:commit(Chain, ConsensusMembers, [Txn])),
?assertEqual(SrcBalance - Amount, t_chain:get_balance(Chain, Src)),
?assertEqual(DstBalance + Amount, t_chain:get_balance(Chain, Dst)),
```

for more transparency and control, the above could start with:

```erlang
Src = t_user:new(),
SrcBalance = 5000,
Dst = t_user:new(),
DstBalance = 0,
Cfg = t_chain:start(Cfg0, #{users_with_hnt => [{Src, SrcBalance}]}),
...
```

Although, it may actually be better for `start` to accept a list of `{Address, Balance}` pairs...

### CODE REVIEW

File: test/blockchain_ct_utils.erl [hunk 1]
```diff
@@ -169,8 +170,8 @@ get_config(Arg, Default) ->
 random_n(N, List) ->
     lists:sublist(shuffle(List), N).
 
-shuffle(List) ->
-    [x || {_,x} <- lists:sort([{rand:uniform(), N} || N <- List])].
+shuffle(Xs) ->
```
Comment (evanmcc) [inline test/blockchain_ct_utils.erl:173]: should we just use the shuffle in `blockchain_utils` instead?

**Review by evanmcc — COMMENTED** [review]
can you squash this and clean up the debug stuff?  sorry for the review delay, this is looking pretty close.

currently isn't compiling locally, but I think that a rebase will fix that.

File: src/blockchain_worker.erl [hunk 2]
```diff
@@ -19,6 +19,7 @@
     num_consensus_members/0,
     consensus_addrs/0,
     integrate_genesis_block/1,
+    integrate_genesis_block__call/1,
```
[the remark this answers is not in the document]
Comment (xandkar) [reply src/blockchain_worker.erl]: > why do we need this?

So we can initialize a chain without `wait_until`-like things. I renamed this `integrate_genesis_block_synchronously` now.

**Comment by xandkar** [comment]
@evanmcc All cleaned-up! Sorry, I didn't mean for you to wade through all the debug-print filth - didn't have the latest patch pushed :grimacing:

File: src/blockchain.erl [hunk 3]
```diff
@@ -1480,9 +1480,19 @@ check_recent_blocks(Blockchain) ->
...
 
+-type crosscheck_error()
+        :: {ledger_delayed_ledger_lag_too_large, non_neg_integer()}
+        |  {ledger_delayed_ledger_lag_too_small, non_neg_integer()}
+        |  {chain_ledger_height_mismatch, non_neg_integer(), non_neg_integer()}
+        |  {fingerprint_mismatch, [term()]}
+        |  {missing_block, non_neg_integer()}
+        .
+
+-spec crosscheck(blockchain()) -> ok | {error, crosscheck_error()}.
```
Comment (xandkar) [inline src/blockchain.erl:1491]: This should've gone into a separate commit. I got overeager with squashing...

**Review by evanmcc — APPROVED** [review]

**Pull request merged**